### ISSUE TITLE
Sequence of SequenceRecordReader is not supported

### ISSUE THREAD

**Issue by Hadi-H** [deeplearning4j/deeplearning4j#4084]
I would like to have a sequence record reader, in which each element of time series is a sequence itself. In other words, I need a sequence of sequence record readers as an input of the "RecordReaderMultiDataSetIterator" of computation graphs. This type of sequence record readers is not supported by deeplearning4j.
I'd like to help you fix the issue by contributing some code, but would like guidance or help in doing so! Any similar example would be very useful to start the implementation.
Thanks!

**Comment by AlexDBlack** [comment]
As noted in gitter - we don't currently support this use case natively for our data pipelines, though we plan on supporting it in the future.

A consequence of this is that you will need to write your data pipeline (DataSetIterator / MultiDataSetIterator) yourself.

Normal RNN data has 3 dimensions: [minibatch, vectorSize, timeSeriesLength]
And masks (for per-time step masking / variable length) have 2 dimensions: [minibatch, timeSeriesLength]

Your "sequence of sequences" data would need to have 4 dimensions: something like [minibatch, seqOfSeqNumber, vectorSize, timeSeriesLength], with masks having 3 dimensions: [minibatch, seqOfSeqNumber, timeSeriesLength]

You'd need to use preprocessors, that do something similar to our RnnToFeedForwardPreProcessor, to make the data suitable for feeding into a RNN. Again, that's something you'll have to write.

I'm not clear exactly what you want to do after that. Pooling over time? Create a "new" sequence out of the last time step of each input sequence?
Both are possible, though the design is quite different, obviously.

**Comment by Hadi-H** [comment]
Thanks for your reply and the information Alex. 
>You'd need to use preprocessors, that do something similar to our RnnToFeedForwardPreProcessor, to make the data suitable for feeding into a RNN. Again, that's something you'll have to write.
I have created a RnnToRnnPreprocessor class. Is it fine in this way?: https://gist.github.com/Hadi-H/683f0b1a4c2ac57431a3b33144b689ef
>Your "sequence of sequences" data would need to have 4 dimensions: something like [minibatch, seqOfSeqNumber, vectorSize, timeSeriesLength], with masks having 3 dimensions: [minibatch, seqOfSeqNumber, timeSeriesLength]
I have tried to modify the "SequenceRecordReaderDataSetIterator" class to accept a data having 4 dimensions, but it's not very clear for me. Could you please give more details on how to implement this? My input is a sequence of sentences and each sentence is a sequence of word. My SequenceRecordReaderDataSetIterator modification is available here: https://gist.github.com/Hadi-H/d95a08030e7afa88a793ac37a94981ad
In this case, I'm trying to use the NumberedFileInputSplit and each sequence of words (each sentence) is in a file, in which each line has a 100 dimensional word representation. If a file has just 1 line and does not have a 100 diemnsional word representation, I'd like to consider it as end of sequence of sentences. At this point the label file having same file number as feature is representative of the sequence of sentences' label (sequence.size() == 1 && sequence.get(0).size() != 100). The data pipeline is not very clear to me. Please help if you have a simpler/better solution. I really stuck at this point.
>I'm not clear exactly what you want to do after that. Pooling over time? Create a "new" sequence out of the last time step of each input sequence?
So, basically, I'd like to have 2 LSTM layers, and then do a binary classification using a RnnOutputLayer. 

Thank you in advance for your help.

**Comment by AlexDBlack** [comment]
That preprocessor looks, to be blunt, quite wrong. 

- It expects 3d input data, when it needs to accept 4d (sequence of sequences = sequence of 3d data = 4d)
- The tensor along dimensions, permutes etc only apply for the 3d case. You'll need something different for 4d
- The reshaping looks wrong:
    - Javadoc: should be [miniBatchSize,seqOfSeqNumber,layerSize,timeSeriesLength] -> [miniBatchSize * seqOfSeqNumber,layerSize,timeSeriesLength] (reverse for backprop)
    - Actual implementation: should be something like .reshape(shape[0]*shape[1], shape[2], shape[3])

**Comment by Hadi-H** [comment]
Thanks Alex, I have modified the preprocessor in this way. If I get it correctly, it need to be used at the first layer (before feeding data to the first LSTM). Am I right? So, basically, in this way: https://gist.github.com/Hadi-H/137b4a09354e7a176348ebfa8fb22b20
>The tensor along dimensions, permutes etc only apply for the 3d case. You'll need something different for 4d

Could you please give a bit more detail on this. I didn't change this part because I don't know what to do. The input data seems fine, however, I get the following error related to the labels "java.lang.IllegalArgumentException: Labels array numColumns (size(1) = 4) does not match output layer number of outputs (nOut = 2)": https://gist.github.com/Hadi-H/a924bc1a05e32c9ecf785ff65c8c312f

Your help is so appreciated. Thanks a lot

**Comment by Hadi-H** [comment]
I've implemented the data pipeline in a way that the miniBatchSize is equal to number of sentences in the sequence of sequence (the miniBatchSize is variable). So in this way, in the RnnToRnnPreProcessor, I feed sequences to the 2nd LSTM that have the miniBatchSize sequence length, same layer size as the input, and assign 1 to the miniBatchSize. However, I get "Mismatched shapes" error.

This is how the input in "public INDArray preProcess(INDArray input, int miniBatchSize) { ..." look like: https://gist.github.com/Hadi-H/7139fb059eab4ab4ffade721c5ecaff1 

I think if I know how to reshape the input data in a way to just consider the activation function output of last record in each sequence, then I can preprocess the data correctly for the sequence of sequence to be fed to the 2nd LSTM. As an example, in the example inputData available in the posted file, the input has minibatchSize =2, sequenceLength = 6, and lasyerSize = 100. I need to change it to minibatchSize =1, sequenceLength = 2, and lasyerSize = 100. Any idea how to do it?

**Comment by agibsonccc** [comment]
Closing — references a removed/rewritten subsystem (DataVec, NLP/Word2Vec, Model Zoo, Spark, UI, or Arbiter). These have been fundamentally rewritten or removed in the upcoming release. Please open a new issue if you experience similar problems with the new version.
